### ISSUE TITLE
[Hold for payment 2024-02-13] [$500] Private Note - Code block paste in private notes does not work (works fine in native apps)

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#29806]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.85.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1697530399668279

## Action Performed:
1. Open the app
2. Open any user report
3. Send 3 backtick code block eg: ```Hii```
4. Hover and click copy to clipboard
5. Click on header to open user details
6. Click on privates notes-> my notes
7. Paste the copied text and observe that app removes backticks while pasting

## Expected Result:
 App should paste text with backticks in private notes

## Actual Result:
App does not paste text with private notes (works fine only in native apps)

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

 

https://github.com/Expensify/App/assets/93399543/fe21985a-eaf4-459a-b788-30c338892c6b

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/db2f07e0-7046-4f05-9fa6-957755079814

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/5b66b28c-a05d-4afd-b21f-903255894055

https://github.com/Expensify/App/assets/93399543/d26511ec-d51c-462f-8506-fefb5be72563

</details>

<details>
<summary>MacOS: Desktop</summary>

 

https://github.com/Expensify/App/assets/93399543/7212464f-bcdd-42e4-81f0-42bac2828632

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013e368c7fef7e59bc</li>
        <li>Upwork Job ID: 1714338003846836224</li>
        <li>Last Price Increase: 2023-11-07</li>
    </ul>
</details>

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Code block paste doesn't work properly in private notes
### What is the root cause of that problem?
Paste of code block hasn't been handled as in our `Composer` component.
### What changes do you think we should make in order to solve the problem?

We need to add listener to paste event and handle paste of html content as we did in our `Composer` component
https://github.com/Expensify/App/blob/8917f95da167f91b8a2ae0c26b78ad1b4d276786/src/components/Composer/index.js#L250-L256

### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
Updated [comment](https://github.com/Expensify/App/issues/29806#issuecomment-1766975028) 
And also I still can reproduce it in the latest main :-

https://github.com/Expensify/App/assets/38649957/630657f5-d40e-430c-8ba8-d25d96cad97f

**Comment by garrettmknight** [comment]
I was able to repro it on web via Chrome and also on the MacOS app.

**Comment by garrettmknight** [comment]
@ArekChr what do you think of that proposal?

**Comment by ArekChr** [comment]
@FitseTLT, Could you provide more details about where the change should be applied?

**Comment by FitseTLT** [comment]
@ArekChr we need to add this code to `PrivateNotesEditPage` and of course use `withNavigation` HOC on the `PrivateNotesEditPage` component exported.
```

    /**
     * Set pasted text to clipboard
     * @param {String} text
     */
    const paste = useCallback((text) => {
        try {
            document.execCommand('insertText', false, text);
            // Pointer will go out of sight when a large paragraph is pasted on the web. Refocusing the input keeps the cursor in view.
            textInput.current.blur();
            textInput.current.focus();
            // eslint-disable-next-line no-empty
        } catch (e) {}
    }, []);

    /**
     * Manually place the pasted HTML into Composer
     *
     * @param {String} html - pasted HTML
     */
    const handlePastedHTML = useCallback(
        (html) => {
            const parser = new ExpensiMark();
            paste(parser.htmlToMarkdown(html));
        },
        [paste],
    );

    /**
     * Paste the plaintext content into Composer.
     *
     * @param {ClipboardEvent} event
     */
    const handlePastePlainText = useCallback(
        (event) => {
            const plainText = event.clipboardData.getData('text/plain');
            paste(plainText);
        },
        [paste],
    );

    const handlePaste = useCallback(
        (event) => {
            const isFocused = privateNotesInput.current.isFocused();

            if (!isFocused) {
                return;
            }

            if (privateNotesInput.current !== event.target) {
                // To make sure the text input does not capture paste events from other inputs, we check where the event originated
                // If it did originate in another input, we return early to prevent the text input from handling the paste
                const isTargetInput = event.target.nodeName === 'INPUT' || event.target.nodeName === 'TEXTAREA' || event.target.contentEditable === 'true';
                if (isTargetInput) {
                    return;
                }

                privateNotesInput.current.focus();
            }

            event.preventDefault();

            const {types} = event.clipboardData;
            const TEXT_HTML = 'text/html';

            // If paste contains HTML
            if (types.includes(TEXT_HTML)) {
                const pastedHTML = event.clipboardData.getData(TEXT_HTML);

                const domparser = new DOMParser();
                const embeddedImages = domparser.parseFromString(pastedHTML, TEXT_HTML).images;

                // Exclude parsing img tags in the HTML, as fetching the image via fetch triggers a connect-src Content-Security-Policy error.
                if (embeddedImages.length > 0 && embeddedImages[0].src) {
                    // If HTML has emoji, then treat this as plain text.
                    if (embeddedImages[0].dataset && embeddedImages[0].dataset.stringifyType === 'emoji') {
                        handlePastePlainText(event);
                        return;
                    }
                }
                handlePastedHTML(pastedHTML);
                return;
            }
            handlePastePlainText(event);
        },
        [handlePastedHTML, handlePastePlainText],
    );

    useEffect(() => {
        // we need to handle listeners on navigation focus/blur as Composer is not unmounting
        // when navigating away to different report
        const unsubscribeFocus = navigation.addListener('focus', () => document.addEventListener('paste', handlePaste));
        const unsubscribeBlur = navigation.addListener('blur', () => document.removeEventListener('paste', handlePaste));

        if (privateNotesInput.current) {
            document.addEventListener('paste', handlePaste);
        }

        return () => {
            unsubscribeFocus();
            unsubscribeBlur();
            document.removeEventListener('paste', handlePaste);
            // eslint-disable-next-line es/no-optional-chaining
            if (!privateNotesInput.current) {
                return;
            }
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

```

**Comment by ArekChr** [comment]
@FitseTLT, I'm a bit confused. Previously, you mentioned that we don't handle pasted HTML in the `Composer`, but now you're suggesting we should add it to the `PrivateNotesEditPage`. Private Notes, does not use the Composer component. I'm unsure about the differences in the code you provided, it seems that is similar to the code in `Composer`. Could you please share the specific code changes and provide more details, especially regarding the exact file where these changes should be applied?

**Comment by ArekChr** [comment]
@FitseTLT, I'm a bit confused. Previously, you mentioned that we don't handle pasted HTML like in the Composer, but now you suggest adding it to the PrivateNotesEditPage. 
I'm unsure about the differences in the code you provided. It is similar to the code in Composer. Please update your proposal with the specific code changes and provide more details, especially regarding the exact file where these changes should be applied?

If you suggest using the same code from the `Composer`, we should create reusable functionality instead of copying it to a different place.

**Comment by FitseTLT** [comment]
> @FitseTLT, I'm a bit confused. Previously, you mentioned that we don't handle pasted HTML like in the Composer, but now you suggest adding it to the PrivateNotesEditPage. I'm unsure about the differences in the code you provided. It is similar to the code in Composer. Please update your proposal with the specific code changes and provide more details, especially regarding the exact file where these changes should be applied, by linking it in the proposal.
> 
> If you suggest using the same code from the `Composer`, we should create a reusable functionality instead of copying it to a different place.

Yes, reusable code, that's exactly what I proposed under `What alternative solutions did you explore? (Optional)`  @ArekChr .
Now I have updated my [proposal](https://github.com/Expensify/App/issues/29806#issuecomment-1766975028) with a working hook that can be used in both places. Thx!!

**Comment by FitseTLT** [comment]
@ArekChr did you check my updated [proposal](https://github.com/Expensify/App/issues/29806#issuecomment-1766975028)? I updated it with the custom hook implementation and result video but if it is better I can also implement the HOC implementation 👍 .

**Comment by FitseTLT** [comment]
Updated [comment](https://github.com/Expensify/App/issues/29806#issuecomment-1766975028)

**Comment by ArekChr** [comment]
@FitseTLT I’m going to check your proposal again soon. I just need a bit more time to look into how code blocks can be pasted on web platforms to make sure your solution works properly.

**Comment by garrettmknight** [comment]
@ArekChr will you be able to review this week?

**Comment by ArekChr** [comment]
Hello, I am wrapping up some urgent matters today and I will be OOO from tomorrow to Monday. Can we assign another C+ here so as not to block the task?

**Comment by garrettmknight** [comment]
Hey @Santhosh-Sellavel - we have a proposal ready for review on this one: https://github.com/Expensify/App/issues/29806#issuecomment-1780831127

**Comment by Santhosh-Sellavel** [comment]
@garrettmknight I'm just gonna unassign this to focus on backlogs.

**Comment by Santhosh-Sellavel** [comment]
Please assign a new C+ here

**Comment by garrettmknight** [comment]
Hey @robertKozik - we have a proposal ready for review on this one: https://github.com/Expensify/App/issues/29806#issuecomment-1780831127

**Comment by yh-0218** [comment]
And I found one issue here.
Current private note is not same with draft note of edit page 
You can check on video

**Comment by garrettmknight** [comment]
@robertKozik can you give this one a look when you get a chance?

**Comment by robertKozik** [comment]
Hi all! sorry for keep you waiting for so long. I've get familiar with the issue a little and I think that the root cause here is that `TextInput` on the Private Notes screen doesn't support pasting HTML content (like it is in `Composer` example). 

After getting acquainted with both of the avaiable proposal, I can see flaws on both of them, which I want to point uout separetly:
- @yh-0218: your proposal is getting rid of not only the issue, but the correctly implemented part of the code as well. We should still be able to copy/paste the HTML content, not only as string. The `Clipboard.canSetHtml` function, which you want to get rid of, is exacly for this purpose - enable the HTML copying
- @FitseTLT - your solution would work, but it would require from us to create platform-dependant versions of the edit private notes screen as we are creating listeners on `document` object. In my opinion it's not worth it in this case to condemn contributors to maintain both very similar versions of code. But **I like more your alternative solution** to pass the markdown version of content as `text/plain`. If you could update your proposal so this one would be the primary solution I would be glad.

I believe that the secondary solution from the @FitseTLT is the most promising. From mine point of view copying function should copy the contant ideally with formatting, so exposing both versions HTML and Markdown of the content is expected here

**Comment by FitseTLT** [comment]
@robertKozik Updated [Proposal](https://github.com/Expensify/App/issues/29806#issuecomment-1766975028) according to your suggestion 👍

**Comment by robertKozik** [comment]
Thanks @FitseTLT for updating. I believe it should be the way to go, but I'm open to discussion about this.

🎀 👀 🎀 C+ reviewed

**Comment by flodnv** [comment]
Hmmm, it seems we are saying that native has the correct behavior whilst web does not. I am not sure I agree with this opinion. When I copy the text from `test` in any app, including here, "test" gets copied to my clipboard, not \`test\`.

What do you think @garrettmknight @robertKozik ?

**Comment by FitseTLT** [comment]
@flodnv Copying to clipboard from a report action is specifically designed to support both plain text and html, here: 
https://github.com/Expensify/App/blob/73a3e7261b28ad9def8dca4e6c9dac238cd0821e/src/pages/home/report/ContextMenu/ContextMenuActions.js#L273-L292
So if you select text from `test` from a report action still it will copy "test" because it prioritize selection over the report action message :
https://github.com/Expensify/App/blob/73a3e7261b28ad9def8dca4e6c9dac238cd0821e/src/pages/home/report/ContextMenu/ContextMenuActions.js#L273
But if you copy from the context menu without selecting any text it will copy `test` (the code block).
So according to our design (code) the expected behaviour when a user copies a report action message (that is a code block) from our context menu is to copy the code block to clipboard, this is a specific feature already incorporated in the app so that users can copy the exact report action message (whether it is code block or simple plain text) so don't confuse it with the common/default copying mechanism that exists in other apps (including github). 👍

**Comment by robertKozik** [comment]
The behavior of copy/pasting via selection (similar to GitHub) will remain unchanged, and the content won't have the styling. My rationale was that copying via the action menu should preserve both content and styling.

Currently, copying a message via the action menu will only maintain markdown styling when pasted into an input adapted for pasting HTML content. Even if we were to adapt our input for HTML, pasting into another app may not always preserve the styling. In current implementation we are assuming that web platform should be able to set/get HTML, which is not always the true. 

If we would not agree with the copying markdown representation of the content I would suggest creating the special TextInput component (something like `HtmlAwareInput`) which would be able to handle pasting HTML. On a downside we will be sentenced to maintain this kind of the component, which would have separate implementation for web and native.

**Comment by flodnv** [comment]
>So according to our design (code) the expected behaviour

Indeed, I am challenging the expected behavior (current design). 

>So if you select text from test from a report action still it will copy "test" because it prioritize selection over the report action message :
>The behavior of copy/pasting via selection (similar to GitHub) will remain unchanged, and the content won't have the styling.

Actually I just tested and it copied `test` (the code blocker). So, is that a bug?

**Comment by FitseTLT** [comment]
> Actually I just tested and it copied test (the code block). So, is that a bug?
Ok. You 're right it copies the code block. 👍 

> Indeed, I am challenging the expected behavior (current design).
I think we should choose the more inclusive markdown over the plain text. And also as a user copies from the report action context menu he expects to have a content similar (both content and styling) to the report action content.

**Comment by robertKozik** [comment]
> I think we should choose the more inclusive markdown over the plain text. And also as a user copies from the report action context menu he expects to have a content similar (both content and styling) to the report action content on pasting.

If that's the case copying the markdown would be inline with the selection copy in newDot

**Comment by flodnv** [comment]
Ok, I asked around and I'm now partly convinced that this is a change we want to make. I think `Actually I just tested and it copied test (the code block). So, is that a bug?` may be a bug but this is out of scope of this issue. 

Quick question about the proposal: 

```
                    if (!Clipboard.canSetHtml()) {
                        Clipboard.setString(markdown);
                    } else {
                        Clipboard.setHtml(content, markdown);
                    }
```

Can we not remove this if/else and do only `Clipboard.setString(markdown);`?

**Comment by FitseTLT** [comment]
> Ok, I asked around and I'm now partly convinced that this is a change we want to make. I think `Actually I just tested and it copied test (the code block). So, is that a bug?` may be a bug but this is out of scope of this issue.
> 
> Quick question about the proposal:
> 
> ```
>                     if (!Clipboard.canSetHtml()) {
>                         Clipboard.setString(markdown);
>                     } else {
>                         Clipboard.setHtml(content, markdown);
>                     }
> ```
> 
> Can we not remove this if/else and do only `Clipboard.setString(markdown);`?

If we remove we will no more be providing html content to the clipboard which is needed in the `handlePasteHtml` implementation in `Compose` component (web), that's why we chose to provide markdown as plain text for other implementation (to also fix this issue) and provide the html for the `Compose` web implementation 👍 .

**Comment by garrettmknight** [comment]
@robertKozik what do you think?

**Comment by flodnv** [comment]
I assigned @FitseTLT, not sure why Melvin did not comment

**Comment by flodnv** [comment]
PR was reverted: https://github.com/Expensify/App/pull/31852

**Comment by FitseTLT** [comment]
@flodnv When I create a PR should I link it to this issue or the other ??

**Comment by flodnv** [comment]
This one; we'll close the other one soon

**Comment by robertKozik** [comment]
@FitseTLT Could you remind me what is the status of regression fix?

**Comment by FitseTLT** [comment]
@robertKozik will get a pr ready for tomorrow 👍

**Comment by garrettmknight** [comment]
@FitseTLT any update on that PR?

**Comment by FitseTLT** [comment]
Sorry for the lateness :pray: PR ready.

**Comment by Santhosh-Sellavel** [comment]
@robertKozik Can you review [this](https://github.com/Expensify/App/pull/32971) seems it is waiting for you?

**Comment by FitseTLT** [comment]
@flodnv @garrettmknight The C+ membered assigned before has now joined Wave and not C+ anymore. So why don't we assign another C+ like @Santhosh-Sellavel , I mean the issue already has the budget for a C+ member, why don't we utilize that? 👍

**Comment by situchan** [comment]
If https://github.com/Expensify/App/issues/29806#issuecomment-1787621999 is still valid, I can take over.

**Comment by FitseTLT** [comment]
bump @garrettmknight @flodnv

**Comment by flodnv** [comment]
Thanks for the bump, done. Let me know if I misunderstood something. All yours @Santhosh-Sellavel 👍

**Comment by FitseTLT** [comment]
@Santhosh-Sellavel will let u know when to review as the pr is too old will have to merge main and test

**Comment by garrettmknight** [comment]
PR is being actively reveiwed.

**Comment by garrettmknight** [comment]
**Summarizing payments for this issue:**

**Reporter:** @dhanashree-sawant $50 paid via Upwork
**Contributor:** @FitseTLT $250 (50% for the regression) paid via Upwork
**Reviewer:** @Santhosh-Sellavel $500

**Comment by garrettmknight** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

 [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR:
 [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
 [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
 [@Santhosh-Sellavel] Determine if we should create a regression test for this bug.
 [@Santhosh-Sellavel] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.
 [@garrettmknight] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by garrettmknight** [comment]
@Santhosh-Sellavel please fill out the checklist when you get a chance and request payment. Dropping to weekly till then.

**Comment by garrettmknight** [comment]
Just waiting on checklist and payment.

**Comment by garrettmknight** [comment]
@Santhosh-Sellavel can you request payment when you get a chance? The checklist was completed in the original issue.

**Comment by Santhosh-Sellavel** [comment]
Feel free to close this one, I'll request later

**Comment by Santhosh-Sellavel** [comment]
Requested on ND

**Comment by JmillsExpensify** [comment]
$500 approved for @Santhosh-Sellavel
